build-tools/packages/build-tools/src/common/tsCompile.ts

```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type * as tsTypes from "typescript";
+import path from "path";
+import { getTscUtils } from "./tscUtils.js";
+
+// Any paths given by typescript will be normalized to forward slashes.
+// Local paths should be normalized to make any comparisons.
+const directorySeparator = "/";
+const backslashRegExp = /\\/g;
+function normalizeSlashes(path: string): string {
+	return path.includes("\\") ? path.replace(backslashRegExp, directorySeparator) : path;
+}
+
+export function tsCompile({
+	command,
+	cwd,
+	packageJsonTypeOverride,
+}: {
+	command: string;
+	cwd: string;
+	packageJsonTypeOverride?: "commonjs" | "module";
+}): number {
+	// Load the typescript version that is in the cwd scope
+	const tscUtils = getTscUtils(cwd);
+
+	const ts = tscUtils.tsLib;
+	let commandLine = tscUtils.parseCommandLine(command);
+	const diagnostics: tsTypes.Diagnostic[] = [];
+	if (commandLine) {
+		const configFileName = tscUtils.findConfigFile(cwd, commandLine);
+		if (configFileName) {
+			commandLine = ts.getParsedCommandLineOfConfigFile(configFileName, commandLine.options, {
+				...ts.sys,
+				getCurrentDirectory: () => cwd,
+				onUnRecoverableConfigFileDiagnostic: (diagnostic: tsTypes.Diagnostic) => {
+					diagnostics.push(diagnostic);
+				},
+			});
+		} else {
+			throw new Error("Unknown config file in command line");
+		}
+	}
+
+	let code = ts.ExitStatus.DiagnosticsPresent_OutputsSkipped;
+	if (commandLine && diagnostics.length === 0) {
+		const incremental = !!(commandLine.options.incremental || commandLine.options.composite);
+
+		const host = incremental
+			? ts.createIncrementalCompilerHost(commandLine.options)
+			: ts.createCompilerHost(commandLine.options);
+		// When specified overrides current directory's package.json type field so tsc may cleanly
+		// transpile .ts files to CommonJS or ESM using compilerOptions.module Node16 or NodeNext.
+		if (packageJsonTypeOverride) {
+			const origReadFile = host.readFile;
+			const packageJsonPath = normalizeSlashes(path.join(cwd, "package.json"));
+			host.readFile = (fileName: string) => {
+				const rawFile = origReadFile(fileName);
+				if (fileName === packageJsonPath && rawFile !== undefined) {
+					// Reading local package.json: override type field
+					const packageJson = JSON.parse(rawFile);
+					return JSON.stringify({ ...packageJson, type: packageJsonTypeOverride });
+				}
+				return rawFile;
+			};
+		}
```

Comment:
Out of paranoia, you could assert/warn if a packageJson was never read.  (Might not be worth it.  Up to you.)

Reply:
Good idea a warning could help us convert.
This does mean it will warn if not useful, which I could see happening if we were trying to be very explicit with all commands. We can adjust. :)